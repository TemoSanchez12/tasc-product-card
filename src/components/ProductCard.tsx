// from react
import React from 'react';
import { createContext } from 'react';

// custom hooks
import useProduct from '../hooks/useProduct';

// styles
import styles from '../styles/styles.module.css';

// interfaces
import {
  ProductContextProps,
  Product,
  InitialValues,
  ProductCartHandlers,
} from '../interfaces/interfaces';

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[]
  children: (args: ProductCartHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  value?: number;
  initialValues?: InitialValues;
}

// product context
export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

// Principal component
export const ProductCard = ({
  value,
  product,
  className,
  style,
  children,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          reset,
          increaseBy,
        })}
      </div>
    </Provider>
  );
};
