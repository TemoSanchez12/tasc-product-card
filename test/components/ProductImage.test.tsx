import React from 'react';
import renderer from 'react-test-renderer';
import ProductCard, { ProductImage } from '../../src/components';

import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('Debe mostrar el componenete correctamente con la imagen pasada por parametreo', () => {
    const wrapper = renderer.create(<ProductImage img="./coffee-mug.png" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('Debe de mostar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
