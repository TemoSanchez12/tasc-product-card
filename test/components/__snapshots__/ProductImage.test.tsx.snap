// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe de mostar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage Debe mostrar el componenete correctamente con la imagen pasada por parametreo 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="./coffee-mug.png"
/>
`;
